### ISSUE TITLE
Feature request: Scrpyt with SRP6

### ISSUE THREAD

**Issue by volkan-ozdemir** [randombit/botan#1791]
Hi all,

I`m just starting to learn cryptography and I`m trying to implement an SRP based solution for a distributed computing system. I got this concern about botan implementation of SRP:

- It looks like botan implementation only permits SHA family of hash functions for verifier derivation, problem is that attacker can easily bruteforce SHA-2 (or similar type hash functions) and gather weak user passwords in case verifier in the database gets leaked. (Similar stuff happened to Blizzard -> http://opine.me/blizzards-battle-net-hack/)

Honestly, I can`t sleep well at night with 0 protection against verifier leakage, one solution would be letting SRP6 use scrypt based hashing instead.

It looks like this should be possible to implement to botan code, thing is I`m really a novice on cryptography and I don`t want to alter botan code at this stage.

Are my concerns about verifier leakage valid? If so, it would be great if someone could implement scrypt to SRP6.

**Comment by randombit** [comment]
You are correct that if the SRP6 verifier leaks then a brute force attack is possible against the user's password. The choice of SHA here is due to this SRP implementation being primarily written for TLS-SRP, and SHA is what RFC 5054 calls for.

This is easy to work around without any changes to the library. Just prehash the password you provide to SRP6 with scrypt. That is when calling `srp6_client_agree` or `generate_srp6_verifier` instead of passing the password directly, first run scrypt over it (perhaps using a combination of the servername and username as the salt, or derive something from the SRP6 salt), then base64 the scrypt output and use that as the SRP6 password.

It is still bad to leak the verifier since anyone with the verifier can impersonate the server. (Assuming there is no additional use of signatures, etc).

It's unlikely the SRP6 code will be extended in any way. SRP6 is a very old PAKE design and these days many better options (such as SPAKE2+ and OPAQUE) are available, so any time spent on PAKE implementation/improvements would go towards implementing one of those.
